src/miral/wayland_shm.cpp

```diff
@@ -0,0 +1,195 @@
+                    shm_dir = dir;
+                    auto fd = open_shm_file();
+                    if (fd >= 0)
+                        return fd;
+                }
+            }
+
+            // Workaround for filesystems that don't support O_TMPFILE (E.g. phones based on 3.4 or 3.10
+            open_shm_file = []
+            {
+                char template_filename[] = "/dev/shm/mir-buffer-XXXXXX";
+                mir::Fd fd{mkostemp(template_filename, O_CLOEXEC)};
+                if (fd != -1)
+                {
+                    if (unlink(template_filename) < 0)
+                    {
+                        return mir::Fd{};
+                    }
+                }
+                return fd;
+            };
+
+            return open_shm_file();
+        };
+
+    auto fd = open_shm_file();
+
+    if (fd < 0) {
+        BOOST_THROW_EXCEPTION((std::system_error{errno, std::system_category(), "Failed to open shm buffer"}));
+    }
+
+    if (auto const error = posix_fallocate(fd, 0, size))
+    {
+        BOOST_THROW_EXCEPTION((std::system_error{error, std::system_category(), "Failed to allocate shm buffer"}));
+    }
+
+    if ((*data = mmap(nullptr, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0)) == MAP_FAILED)
+    {
+        BOOST_THROW_EXCEPTION((std::system_error{errno, std::system_category(), "Failed to mmap buffer"}));
+    }
+
+    return wl_shm_create_pool(shm, fd, size);
+}
+}
+
+struct miral::tk::WaylandShmPool
+{
+public:
+    WaylandShmPool(wl_shm_pool* shm_pool, void* data, size_t size)
+        : shm_pool{shm_pool},
+          data{data},
+          size{size}
+    {
+    }
+
+    ~WaylandShmPool()
+    {
+        munmap(data, size);
+        wl_shm_pool_destroy(shm_pool);
+    }
+
+    wl_shm_pool* const shm_pool;
+    void* const data;
+    size_t const size;
+};
+
+wl_buffer_listener const miral::tk::WaylandShmBuffer::buffer_listener {handle_release};
+
+miral::tk::WaylandShmBuffer::WaylandShmBuffer(
+    std::shared_ptr<WaylandShmPool>  const& pool,
+    void* data,
+    geom::Size const&size,
+    geom::Stride const& stride,
+    wl_buffer* buffer)
+    : pool{pool},
+      data_{data},
+      size_{size},
+      stride_{stride},
+      buffer{buffer}
+{
+    wl_buffer_add_listener(buffer, &buffer_listener, this);
+}
+
+miral::tk::WaylandShmBuffer::~WaylandShmBuffer()
+{
+    wl_buffer_destroy(buffer);
+}
+
+auto miral::tk::WaylandShmBuffer::use() -> wl_buffer*
+{
+    if (self_ptr)
+    {
+        mir::fatal_error("WaylandShmBuffer used multiple times");
+    }
+    self_ptr = shared_from_this();
+    return buffer;
+}
+
+void miral::tk::WaylandShmBuffer::handle_release(void *data, wl_buffer*)
+{
+    auto const self = static_cast<WaylandShmBuffer*>(data);
+    self->self_ptr.reset();
+}
+
+miral::tk::WaylandShm::WaylandShm(wl_shm* shm)
+    : shm{shm}
+{
+}
+
+auto miral::tk::WaylandShm::get_buffer(geom::Size const& size, geom::Stride const& stride) -> std::shared_ptr<WaylandShmBuffer>
```

Comment:
It seems a bit of a foot-gun to have a method that will allocate buffers if there are none free but with no way to free those buffers other than by asking for a different `size`/`stride` pair.